tools/wavetool.py

```diff
@@ -0,0 +1,150 @@
+            'type': cmd['generate'], 'amp': cmd['amp'][0], 'freq': cmd['freq'][0],
+            'phase': cmd['phase'][0], 'chan': cmd['channel'], 'sample_rate': cmd['sample_rate'],
+            'duration': cmd['duration'][0]
+        }
+        wave_data = generate_sinusoids(**wave_param)
+    else:
+        raise Exception('invalid generate function, will generate nothing')
+    return wave_data
+
+def save_wave(wave_data, **cmd):
+    wave_path = cmd['output']
+    if wave_path == '.' or not wave_path.endswith('wav'):
+        wave_path = wave_path + '/tmp.wav'
+    sample_bits = cmd['bits']
+    np_types = {'S8': np.int8, 'S16': np.int16, 'S32': np.int32}
+    if sample_bits in np_types.keys():
+        # range of wave_data is floating point [-1.0, 1.0]. When saving to any integer format use its full range.
+        wave_data = (np.iinfo(np_types[sample_bits]).max * wave_data).astype(np_types[sample_bits])
+    wavefile.write(wave_path, cmd['sample_rate'], wave_data)
+
+def do_wave_analysis(**cmd):
+    fs_wav, wave = wavefile.read(cmd['recorded_wave'])
+    fs_ref, ref_wave = wavefile.read(cmd['reference_wave'])
+    if fs_wav != fs_ref:
+        print('Can not compare wave with different sample rate')
+        sys.exit(1)
+    if cmd['analyze'] == 'smart_amp':
+        analyze_wav_smart_amp(wave, ref_wave, fs_wav)
+
+# remove digital zeros in two sides
+def trim_wave(wave):
+    # once waves go through DAC/ADC, zero will become small value close to zero,
+    # here we set the digital zero threshold to 100, and cut samples below 100
+    # of two sides, this has no negative impact to processing.
+    DIGITAL_ZERO_THRESHOLD = 100
+    wave_mono = wave[:,0]
+    left_idx = 0
+    right_idx = wave_mono.shape[0] - 1
+    while True:
+        if abs(wave_mono[left_idx]) > DIGITAL_ZERO_THRESHOLD:
+            break
+        left_idx = left_idx + 1
+    while True:
+        if abs(wave_mono[right_idx]) > DIGITAL_ZERO_THRESHOLD:
+            break
+        right_idx = right_idx - 1
+    return wave[left_idx:right_idx,:], left_idx
+
+# float point binary comparison is not supported, and will not be supported
+# check recorded wave through smart amplifier component
+def analyze_wav_smart_amp(wave, ref_wave, fs):
+    trimed_ref_wave, _ = trim_wave(ref_wave)
+    # compare the first two channel
+    trimed_wave_ch_0_1, delay1 = trim_wave(wave[:,0:2])
+    trimed_ref_wave = trimed_ref_wave[0:trimed_wave_ch_0_1.shape[0],:]
+    compare_result0_1 = np.array_equal(trimed_ref_wave, trimed_wave_ch_0_1)
+    # compare the second two channel
+    trimed_wave_ch_2_3, delay2 = trim_wave(wave[:,2:4])
+    trimed_ref_wave = trimed_ref_wave[0:trimed_wave_ch_2_3.shape[0],:]
+    compare_result2_3 = np.array_equal(trimed_ref_wave, trimed_wave_ch_2_3)
+
+    smart_amp_delay = ((delay2 - delay1) / fs * 1000)
+    print('Delay of smart amplifier is %0.3fms' % smart_amp_delay)
+    if compare_result0_1 and compare_result2_3 and smart_amp_delay < 5.:
```

Comment:
Magic number 5 (ms), since it's acceptance criteria could it be defined in the beginning of module and described?